hypothesis-python/src/hypothesis/internal/conjecture/pareto.py

```diff
@@ -0,0 +1,189 @@
+        return DominanceRelation.EQUAL
+
+    if sort_key(right.buffer) < sort_key(left.buffer):
+        result = dominance(right, left)
+        if result == DominanceRelation.LEFT_DOMINATES:
+            return DominanceRelation.RIGHT_DOMINATES
+        else:
+            # Because we have sort_key(left) < sort_key(right) the only options
+            # are that right is better than left or that the two are
+            # incomparable.
+            assert result == DominanceRelation.NO_DOMINANCE
+            return result
+
+    # Either left is better or there is no dominance relationship.
+    assert sort_key(left.buffer) < sort_key(right.buffer)
+
+    # The right is more interesting
+    if left.status < right.status:
+        return DominanceRelation.NO_DOMINANCE
+
+    # Things that are interesting for different reasons are incomparable in
+    # the dominance relationship.
+    if (
+        left.status == Status.INTERESTING
+        and left.interesting_origin != right.interesting_origin
+    ):
+        return DominanceRelation.NO_DOMINANCE
+
+    for target, score in left.target_observations.items():
+        if (
+            target in right.target_observations
+            and right.target_observations[target] > score
+        ):
+            return DominanceRelation.NO_DOMINANCE
+
+    return DominanceRelation.LEFT_DOMINATES
+
+
+class ParetoFront(object):
+    """Maintains an approximate pareto front of ConjectureData objects. That
+    is, we try to maintain a collection of objects such that no element of the
+    collection is pareto dominated by any other. In practice we don't quite
+    manage that, because doing so is computationally very expensive. Instead
+    we maintain a random sample of data objects that are "rarely" dominated by
+    any other element of the collection (roughly, no more than about 10%).
+
+    Only valid test cases are considered to belong to the pareto front - any
+    test case with a status less than valid is discarded.
+    """
+
+    def __init__(self, random):
+        self.__random = random
+        self.__eviction_listeners = []
+
+        self.__front = []
+        self.__contained = {}
+        self.__pending = None
+
+    def add(self, data):
+        """Attempts to add ``data`` to the pareto front. Returns True if this
+        resulted ``data`` being added, otherwise returns False (including if
+        data is already in the collection)."""
+        if data.status < Status.VALID:
+            return False
+
+        if not self.__front:
+            self.__add(data)
+            return True
+
+        if data.buffer in self.__contained:
+            return False
+
+        # We add data to the pareto front by adding it unconditionally and then
+        # doing a certain amount of randomized "clear down" - testing a random
+        # set of elements (currently 10) to see if they are dominated by
+        # something else in the collection. If they are, we remove them.
```

Comment:
I can see that this makes sense in that it's constant-time, and doing anything strictly optimal would be very expensive.

However it seems that with a moderate number of targets, we could have an enormous number of points on the pareto front of which very few would be expected to dominate a point (analogous to the increasing proportion of volume close to the surface of an n-sphere as n increases).

- What memory footprint might `__self.contained` end up with?  Plausible cases seem to me to include keeping almost every example in memory.
- Is it worth running a pass after we finish executing examples to remove examples not on the exact front from the DB?  This would be a slowdown, but not per-example.
- Regardless, it would be nice to have a discussion of the space-performance of this approach in the comment block.

Reply:
I think if this starts to cause us memory usage problems then the solution is to effectively page out to disk - rather than storing test cases in the front, we store IDs (based on a hash of the buffer), and look up test cases by:

1. Looking up the ID in a local cache.
2. On cache miss, reading it off the disk (storing buffers in a directory by ID)
3. Looking up the buffer in the engine using `cached_test_function`

But the size of the pareto front is bounded by the number of valid examples we see, and we don't currently have any workflows that really work very well when `max_examples` is huge anyway.

> Is it worth running a pass after we finish executing examples to remove examples not on the exact front from the DB? This would be a slowdown, but not per-example.

Checking the exact pareto front is intrinsically quadratic, so I think probabilistic bounds are the best we can do.

I don't think the fact that this is an approximate pareto front makes much of a difference to the size (no more than an order of magnitude) - when an example is added typically only about 10% of the pareto front could dominate it (or it wouldn't have been added).

I'll see if I can improve the clear down logic a bit to improve the approximation though.

Reply:
Ah, the order of magnitude thing isn't true. There are scenarios where this can be arbitrarily larger than the pareto front. I think they should be relatively rare though.

Reply:
Yeah, if I'm thinking of plausible large scenarios they mostly have large real fronts, rather than a significant penalty from the approximation.

With the improved cleardown (nice!), I think anything further would get into premature optimisation.